### PR TITLE
gh-124594: Create and reuse one context for the entire asyncio REPL session

### LINKED ISSUE #87585: contextvars does not work properly in asyncio REPL.
BPO | [43419](https://bugs.python.org/issue43419)
--- | :---
Nosy | @asvetlov, @1st1, @lanfon72
PRs | <li>python/cpython#24773</li>

<sup>*Note: these values reflect the state of the issue at the time it was migrated and might not reflect the current state.*</sup>

<details><summary>Show more details</summary><p>

GitHub fields:
```python
assignee = None
closed_at = None
created_at = <Date 2021-03-06.14:09:53.246>
labels = ['type-bug', '3.8', '3.9', 'expert-asyncio']
title = 'contextvars does not work properly in asyncio REPL.'
updated_at = <Date 2021-03-06.15:06:01.651>
user = 'https://github.com/lanfon72'
```

bugs.python.org fields:
```python
activity = <Date 2021-03-06.15:06:01.651>
actor = 'lanfon72'
assignee = 'none'
closed = False
closed_date = None
closer = None
components = ['asyncio']
creation = <Date 2021-03-06.14:09:53.246>
creator = 'lanfon72'
dependencies = []
files = []
hgrepos = []
issue_num = 43419
keywords = ['patch']
message_count = 1.0
messages = ['388199']
nosy_count = 3.0
nosy_names = ['asvetlov', 'yselivanov', 'lanfon72']
pr_nums = ['24773']
priority = 'normal'
resolution = None
stage = 'patch review'
status = 'open'
superseder = None
type = 'behavior'
url = 'https://bugs.python.org/issue43419'
versions = ['Python 3.8', 'Python 3.9']
```

</p></details>

### LINKED ISSUE #124594: Every input in asyncio REPL (wrongly) runs in a separate PEP 567 context
# Bug report

### Bug description:

The bug is twofold, affects the asyncio REPL, and refers to [PEP 567 contexts](https://peps.python.org/pep-0567/#contextvars-context) and [context variables](https://peps.python.org/pep-0567/#contextvars-contextvar).

1. Non-async inputs don't run in the same context.

    <details>

    This is how we can use the standard Python REPL to demonstrate context variables (without asyncio tasks):
    ```
    Python 3.14.0a0 (heads/main-dirty:8447c933da3, Sep 25 2024, 15:45:56) [GCC 11.4.0] on linux
    Type "help", "copyright", "credits" or "license" for more information.
    >>> import contextvars
    >>> var = contextvars.ContextVar("var")
    >>> var.set("ok")
    <Token var=<ContextVar name='var' at ...> at ...>
    >>> var.get()
    'ok'
    ```
  
    I'd expect the asyncio REPL to be a slightly better tool for presenting them though. However, I ran into a problem with running the same sequence of instructions in that REPL, because it implicitly copies the context for every input and then runs the instructions in that copied context:
  
    ```
    asyncio REPL 3.14.0a0 (heads/main-dirty:8447c933da3, Sep 25 2024, 15:45:56) [GCC 11.4.0] on linux
    Use "await" directly instead of "asyncio.run()".
    Type "help", "copyright", "credits" or "license" for more information.
    >>> import asyncio
    >>> import contextvars
    >>> var = contextvars.ContextVar("var")
    >>> var.set("ok")
    <Token var=<ContextVar name='var' at ...> at ...>
    >>> var.get()
    Traceback (most recent call last):
      File ".../cpython/Lib/concurrent/futures/_base.py", line 448, in result
        return self.__get_result()
               ~~~~~~~~~~~~~~~~~^^
      File ".../cpython/Lib/concurrent/futures/_base.py", line 393, in __get_result
        raise self._exception
      File ".../cpython/Lib/asyncio/__main__.py", line 40, in callback
        coro = func()
      File "<python-input-3>", line 1, in <module>
        var.get()
        ~~~~~~~^^
    LookupError: <ContextVar name='var' at ...>
    ```
    </details>

    I assume it to just be a leftover of before [the `context` parameter for asyncio routines became a thing in 3.11](https://github.com/python/cpython/issues/91150).

3. Async inputs don't run in the same context.
    Similar to the first part, but this applies to every coroutine being run in the REPL.
    <details>

    By standard, a simple coroutine _does not_ run in a separate context (in contrast to [tasks](https://docs.python.org/3/library/asyncio-task.html#asyncio.Task) that document their handling of context). Consider the following script:

    ```py
    import asyncio
    from contextvars import ContextVar
    
    var = ContextVar("var", default="unset")
    
    async def setvar() -> None:
        var.set("set")

    async def main() -> None:
        await setvar()  # runs in *this* context
        print(var.get())  # gets us "set", not "unset"
    
    asyncio.run(main())
    print(var.get())  # coro passed to asyncio.run() runs in a copied context, which gets us "unset" here
    ```

    In the REPL, I'd expect that we're inside the "global" context of `asyncio.run()`—what I mean by this is that the context we work in (for the entire REPL session) is the same one that is reused to run every asynchronous input, because we're in the same "main" task, and the point of asyncio REPL is to be able to omit it. While it is expected that every asynchronous input with top-level `await` statements becomes a task for the current event loop internally, I would imagine that the behavior resembles what I'd get by merging all the inputs together and putting them inside a `async def main()` function in a Python script.
    For that reason, I consider the following a part of the bug:

    ```
    asyncio REPL 3.14.0a0 (heads/main-dirty:8447c933da3, Sep 25 2024, 15:45:56) [GCC 11.4.0] on linux
    Use "await" directly instead of "asyncio.run()".
    Type "help", "copyright", "credits" or "license" for more information.
    >>> import asyncio
    >>> import contextvars
    >>> var = contextvars.ContextVar("var")
    >>> async def setvar():
    ...     var.set("ok")
    ...     
    >>> await setvar()
    >>> var.get()
    Traceback (most recent call last):
      File ".../cpython/Lib/concurrent/futures/_base.py", line 448, in result
        return self.__get_result()
               ~~~~~~~~~~~~~~~~~^^
      File ".../cpython/Lib/concurrent/futures/_base.py", line 393, in __get_result
        raise self._exception
      File ".../cpython/Lib/asyncio/__main__.py", line 40, in callback
        coro = func()
      File "<python-input-4>", line 1, in <module>
        var.get()
        ~~~~~~~^^
    LookupError: <ContextVar name='var' at ...>
    ```

    I'd expect every asynchronous input to not run in an implicitly copied context, because that's what feels like if we were writing a huge `async def main()` function—every `await` does not create a task, it just awaits the coroutine (the execution of which is managed by the event loop).

    Instead of comparing to a huge `async def main()`, I could say that we're reusing the same event loop over and over, and that can theoretically mean we want one global context only, not a separate one every time.
    </details>

Interestingly, [IPython](https://github.com/ipython/ipython?tab=readme-ov-file#ipython-productive-interactive-computing) (which works asynchronously under the hood too) only runs into part 2.

Thanks to the already mentioned GH-91150, this is very simple to manage and fix. Expect a PR soon.

### CPython versions tested on:

CPython main branch

### Operating systems tested on:

Linux

### Linked PRs
* gh-124595
* gh-124848
* gh-124849

### PR DESCRIPTION
**Pull request by johnslavik**
Fixes gh-124594.
Closes gh-87585.

I was initially slightly unsure how to test it, and `spawn_repl("-m", "asyncio")` is proven to be quirky (launch `./python -I -i -m asyncio` by yourself and then type in `exit` and see what happens)—please let me know if [there's a better way](https://www.youtube.com/watch?t=566&v=p33CVV29OG8&feature=youtu.be).

I used `copy_context()` to assign the copied context to the `AsyncIOInteractiveConsole.context` attribute, and didn't make a `context` parameter as I didn't see that necessary (it doesn't really seem to be a part of the public API, even though it can be imported and used by anyone, and it's also not in the typeshed).

I believe `copy_context()` makes more sense than manually constructing the context with `Context()` because the `pythonstartup.py` script can set some context variables, but I'm open to discussion about that.

### CODE REVIEW

File: Misc/NEWS.d/next/Library/2024-09-26-13-43-39.gh-issue-124594.peYhsP.rst [hunk 1]
```diff
@@ -0,0 +1 @@
+All asyncio REPL prompts run in the same context
```
Comment (ZeroIntensity) [inline Misc/NEWS.d/next/Library/2024-09-26-13-43-39.gh-issue-124594.peYhsP.rst]:
```suggestion
All :mod:`asyncio` REPL prompts run in the same :class:`contextvars.Context`.
```

**Comment by ZeroIntensity** [comment]
This needs a NEWS entry, per the bot, as this is a user-facing bug. (Also, hi!)

**Comment by Eclips4** [comment]
Is this specific to a new REPL, or is it also a problem with the old REPL? If so, there's no need to backport it to 3.12.

**Comment by ZeroIntensity** [comment]
> Is this specific to a new REPL, or is it also a problem with the old REPL? If so, there's no need to backport it to 3.12.

Oh, I didn't check. Does the asyncio REPL use the new REPL?

**Comment by ZeroIntensity** [comment]
> This needs a NEWS entry, per the bot, as this is a user-facing bug. (Also, hi!)

Nevermind, I was being dumb, there's already a news entry. Ignore that!

**Comment by Eclips4** [comment]
> > Is this specific to a new REPL, or is it also a problem with the old REPL? If so, there's no need to backport it to 3.12.
> 
> Oh, I didn't check. Does the asyncio REPL use the new REPL?

Yes

**Comment by bswck** [comment]
Another fun fact I've also observed is that `user_inputX = ...` variables used in tests are not in fact atomic user inputs as opposed to the standard REPL's tests, and every line break in them causes an immediate prompt EOF (try to indent the body of the `set_var()` function and see what happens). I'm not sure how to handle that; it seems to be a part of a separate bug triggered by `-i` and `-m asyncio` used together.

**Review by ZeroIntensity — APPROVED** [review]
I manually confirmed locally that this fixes the problem. FWIW, @Eclips4, I checked that this does apply to 3.12, so we're fine with the backport.

**Comment by kumaraditya303** [comment]
> I believe copy_context() makes more sense than manually constructing the context with Context() because the pythonstartup.py script can set some context variables, but I'm open to discussion about that.

I haven't followed the new repl change so can't review but `copy_context` doesn't allows back propagation across threads, not sure if that matters here.

**Comment by ZeroIntensity** [comment]
As far as I can see, this isn't specific to the new REPL.

**Review by asvetlov — APPROVED** [review]

**Comment by bswck** [comment]
@asvetlov Thanks! This also closes out gh-87585 and gh-24773.

**Pull request merged**